trace_replay/block_cache_tracer.cc

```diff
@@ -215,4 +212,47 @@ Status BlockCacheTraceReader::ReadAccess(BlockCacheTraceRecord* record) {
   return Status::OK();
 }
 
+AtomicBlockCacheTraceWriter::AtomicBlockCacheTraceWriter() {
+  writer_.store(nullptr);
+}
+
+AtomicBlockCacheTraceWriter::~AtomicBlockCacheTraceWriter() { EndTrace(); }
+
+Status AtomicBlockCacheTraceWriter::StartTrace(
+    Env* env, const TraceOptions& trace_options,
+    std::unique_ptr<TraceWriter>&& trace_writer) {
+  InstrumentedMutexLock lock_guard(&writer_mutext_);
+  if (writer_.load()) {
+    return Status::OK();
+  }
+  trace_options_ = trace_options;
+  writer_.store(
+      new BlockCacheTraceWriter(env, trace_options, std::move(trace_writer)));
+  return writer_.load()->WriteHeader();
+}
+
+void AtomicBlockCacheTraceWriter::EndTrace() {
+  InstrumentedMutexLock lock_guard(&writer_mutext_);
+  if (!writer_.load()) {
+    return;
+  }
+  delete writer_.load();
+  writer_.store(nullptr);
+}
+
+Status AtomicBlockCacheTraceWriter::WriteBlockAccess(
+    const BlockCacheTraceRecord& record) {
+  if (!writer_.load()) {
+    return Status::OK();
+  }
+  if (!ShouldTrace(record, trace_options_)) {
+    return Status::OK();
+  }
+  InstrumentedMutexLock lock_guard(&writer_mutext_);
```

Comment:
Can you move `lock_guard` right before `return` ?

Reply:
I put it after the return here because if we sample 1% of blocks, we will acquire this writer_mutex _ for 1% of the time.